lib/Dialect/TensorExt/Transforms/ImplementShiftNetwork.h

```diff
@@ -10,6 +16,216 @@ namespace tensor_ext {
 #define GEN_PASS_DECL_IMPLEMENTSHIFTNETWORK
 #include "lib/Dialect/TensorExt/Transforms/Passes.h.inc"
 
+// A permutation of 0..n-1. This vector should always have size n and contain
+// each integer from 0 to n-1 exactly once.
+using Permutation = FrozenVector<int64_t>;
+
+// A group of indices to rotate together
+using RotationGroup = DenseSet<int64_t>;
+
+// Convert an input->output index mapping to a canonical left-shift amount for
+// a given tensor size.
+// Example: 1 -> 13 with a 64-size tensor should produce a rotation of 52
+// Example: 13 -> 1 with a 64-size tensor should produce a rotation of 12
+inline int64_t normalizeShift(int64_t input, int64_t output,
+                              int64_t tensorSize) {
+  int64_t shift = (output - input) % tensorSize;
+  shift = -shift;  // Account for leftward rotations
+  if (shift < 0) {
+    shift += tensorSize;
+  }
+  return shift;
+}
+
+// The ShiftStrategy class applies power-of-two shifts to each set bit in
+// LSB-to-MSB order, 1, 2, 4, 8, .... Each shift amount is considered a "round"
+// in which a group of indices are attempted to be shifted together. This can be
+// used both to identify conflicts for the graph coloring technique of
+// Vos-Vos-Erkin, and also to construct the concrete shift network after a
+// partition has been decided by Vos-Vos-Erkin.
+struct ShiftRound {
+  // Maps the index of the original input to its current position in the
+  // tensor. This may contain multiple indices mapping to the same slot due to
+  // conflicts in the shifting strategy.
+  SmallVector<int64_t> positions;
+  // The set of indices that are rotated left in this round. This can be used
+  // to generate a mask to select the indices that need rotating.
+  SmallVector<int64_t> rotatedIndices;
+  // The amount rotated left;
+  int64_t rotationAmount;
+};
+
+class ShiftStrategy {
+ public:
+  ShiftStrategy() = default;
+
+  SmallVector<ShiftRound> getRounds() const { return rounds; }
+
+  // Run the shifting strategy and populate the `rounds` member variable.
+  void evaluate(const Permutation &permutation, const RotationGroup &group) {
```

Comment:
Would it be possible to just put the declarations in the header file and the implementations still in the `cpp` file? It's not a blocker but I feel it's a better organizational setup, and avoids issues like importing this header file overwriting the `DEBUG_TYPE`.